### ISSUE TITLE
Memory surge when used delayed+map_blocks

### ISSUE THREAD

**Issue by chrisroat** [dask/dask#6828]
**What happened**:

I am porting some skimage functionality to dask for histogram matching.  The code seems to work correctly for small examples and unit tests, so I believe it is coded correctly (famous last words, I know).  However, when performing a moderate-sized task, the dask version is goes slow and uses a lot of memory.  The numpy version operates with no such issues on the same data.

**What you expected to happen**:

On a single-chunk moderate example, numpy and dask should give the same performance.  Likely, there is some subtlety to dask I do not appreciate here.

**Minimal Complete Verifiable Example**:

```python

# Numpy code
# Copied for skimage.exposure.histogram_matching.py for comparison
def np_match(source, template, dtype=np.uint16):
    _, src_unique_indices, src_counts = np.unique(source.ravel(), return_inverse=True, return_counts=True)

    tmpl_values, tmpl_counts = np.unique(template.ravel(), return_counts=True)

    src_quantiles = np.cumsum(src_counts) / source.size
    tmpl_quantiles = np.cumsum(tmpl_counts) / template.size

    interp_a_values = np.interp(src_quantiles, tmpl_quantiles, tmpl_values).astype(dtype)

    return interp_a_values[src_unique_indices].reshape(source.shape)

```

```
# Dask port

def _interpolate(src_quantiles, tmpl_quantiles, tmpl_values, dtype):
    return np.interp(src_quantiles, tmpl_quantiles, tmpl_values).astype(dtype)

def da_match(source, template, dtype=np.uint16):
    _, src_unique_indices, src_counts = da.unique(source.ravel(), return_inverse=True, return_counts=True)

    tmpl_values, tmpl_counts = da.unique(template.ravel(), return_counts=True)

    src_quantiles = da.cumsum(src_counts) / source.size
    tmpl_quantiles = da.cumsum(tmpl_counts) / template.size

    # Debugging here and computing any of the above variables is fast, as expected.  It's this next step,
    # passing the delayed into the map_blocks, which seems to trigger an issue.
    # (Yes, I do understand this is a bottleneck in the graph, and must only use int16 or smaller dtypes.)
    interp_a_values = dask.delayed(_interpolate)(src_quantiles, tmpl_quantiles, tmpl_values, dtype)
    interp_a_values = da.from_delayed(interp_a_values, dtype=dtype, shape=(np.nan, ))  # shape is src_quantiles.shape
    
    def lookup(chunk, values):
        return values[chunk]
    
    result = src_unique_indices.map_blocks(lookup, values=interp_a_values, dtype=interp_a_values.dtype)
    
    return result.reshape(source.shape)
```

Test case:
```
chn0 = np.random.randint(0, 500, (63, 300, 300), dtype=np.uint16)
chn1 = np.random.randint(0, 500, (63, 300, 300), dtype=np.uint16)

result_np = np_match(chn0, chn1)  # sub-second, no noticable memory usage

result_da = da_match(da.from_array(chn0), da.from_array(chn1)).compute()  # 11 seconds, 50GB
```

**Anything else we need to know?**:

I have experimented with passing a delayed shape to `from_delayed` and using the meta keyword in `map_blocks`.

**Environment**:

- Dask version: 2.30.0
- Python version: 3.8.5
- Operating System: Ubuntu 20.4
- Install method (conda, pip, source): pip

**Comment by jsignell** [comment]
I can reproduce this locally and I think the issue is actually with `da.unique`. I tried to compute `unique_indices` and was able to kill my workers. I think `return_inverse` might be causing the issue here. 

```python
chn0 = np.random.randint(0, 500, (63, 300, 300), dtype=np.uint16)
source = da.from_array(chn0)

_, src_unique_indices, src_counts = da.unique(source.ravel(), return_inverse=True, return_counts=True)
src_unique_indices.compute()
```

**Comment by jsignell** [comment]
Pinging @jakirkham on this since it seems like you implemented `da.unique`.

**Comment by chrisroat** [comment]
Thanks for finding the root issue.  I'm still trying to wrap my head around the code, but here's a better minimum example.  The graph, below, isn't complex.

```python
import dask
import dask.array as da
arr = da.random.randint(500, size=10**6)
_, indices = da.unique(arr, return_inverse=True)
indices_c = indices.compute()
```

![image](https://user-images.githubusercontent.com/1053153/98849558-43288780-2408-11eb-8927-84bfb6ef4201.png)

**Comment by jakirkham** [comment]
It's worth noting that `Delayed` forces all chunks into a single chunk. If that single chunk doesn't fit in memory, that would potentially cause issues. Is it possible to avoid using `Delayed` and do something else?

**Comment by chrisroat** [comment]
That `delayed` call is intended to fit in memory.  It works as long as an array of unique values fits in memory (which is fine up to uint16).  I have checks in my actual code to enforce this.  It's a bottleneck, but I believe it is necessary.

The culprit for the memory use looks to be the following during the inverse calculation.  The `ar` variable is the incoming array, and the `out["values"]` is the list of unique elements. So it creates a size (num_orig_array_elements, num_unique_elements) array of int64.  For the example with 1M elements using 500 values, that's few GB.
https://github.com/dask/dask/blob/3b1303f83a463549fac74a0657e7b67395d5ba12/dask/array/routines.py#L1058-L1059

I don't know if there is a different approach, but will think on it.

**Comment by jsignell** [comment]
There is a comment in the `unique` code that it could be rewritten using a tree reduction approach. Maybe now's the time?

**Comment by jakirkham** [comment]
Yeah that's tracked as issue ( https://github.com/dask/dask/issues/2851 ). It's certainly a doable thing if someone has interest in doing that.

I thought about bring it up. However as we are putting all of the data into a single chunk, which `unique` is running on, I don't think it would help.

This actually raises another question @chrisroat. Is there any reason not to just call `map_blocks` (or perhaps use `delayed`) with `np.unique` in this case? It seems like `da.unique` doesn't offer much for this use case that can't be easily accomplished some other way.

**Comment by chrisroat** [comment]
In this case, the `unique` call is over all the data, not just per-block.   It's the interpolation being done on a single chunk via delayed.

I'll take a look at the tree reduction idea.

**Comment by chrisroat** [comment]
I took a peak at how this might work, and have to say that I don't fully follow all the meta and reduction code.

For the moment, I have the following which seems to be working, though not optimized or vetted on large data.  It's pieced together from some ideas on this and related threads.  I didn't figure out how to pass unknown-shapes around without triggering problems, so it conservatively passes around fixed-size histograms of size 256 or 65536.
https://gist.github.com/chrisroat/7314ba14eeb28390349a9c9d97477b91

I did make some attempts to fit it into the current reductions/routines....

I tried first ignoring the return_xxx arguments, and seeing how we might approach a super simple unique via [reduction](https://github.com/dask/dask/blob/061613eb8ef983e6bc1d620ed43634d8a4f6e1af/dask/array/reductions.py#L49).  It straightforward to use `np.unique` (wrapped for `keepdims`) as the `chunk` and `aggregate` arguments.  

I then tried the next level of complexity, using a structured array with "values" and "counts".  I believe this is where @jakirkham was going with the `_unique_internal` method.  It looks like:

```python
def combine_unique(arr, axis=None, keepdims=None):
    return _unique_internal(arr["values"], None, arr["counts"])

def unique(a, split_every=None):
    a = a.ravel()
    dt = [("values", a.dtype), ("counts", np.intp)]
    arr = np.empty(a.shape, dtype=dt)
    arr["values"] = a
    arr["counts"] = da.ones_like(a, dtype=np.intp)
    result = reduction(
        arr,
        combine_unique,
        combine_unique,
        dtype=arr.dtype,
        split_every=split_every,
    )
    # TODO: unpack result in the end.
    return result
```
The problem I run into is that in the reduction, I bump into `TypeError: cannot perform reduce with flexible type` here:
https://github.com/dask/dask/blob/061613eb8ef983e6bc1d620ed43634d8a4f6e1af/dask/array/reductions.py#L331

Regarding the inverse mapping, I presume that is done as a subsequent map_blocks after the values list has been created (as in my gist above).

**Comment by jakirkham** [comment]
Basically it is complaining that structured `dtype`s can't be reduced

```python
In [1]: import numpy as np

In [2]: a = np.array([(0, 1.0)], dtype=[("a", int), ("b", float)])

In [3]: a.sum()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-393d3db239ee> in <module>
----> 1 a.sum()

~/miniconda/envs/dask/lib/python3.8/site-packages/numpy/core/_methods.py in _sum(a, axis, dtype, out, keepdims, initial, where)
     45 def _sum(a, axis=None, dtype=None, out=None, keepdims=False,
     46          initial=_NoValue, where=True):
---> 47     return umr_sum(a, axis, dtype, out, keepdims, initial, where)
     48 
     49 def _prod(a, axis=None, dtype=None, out=None, keepdims=False,

TypeError: cannot perform reduce with flexible type
```

I think structured `dtype`s are unique to NumPy. It doesn't seem possible to create them with CuPy and I doubt Sparse supports them. Would suggest just special casing this when they show up. Can do a check like `meta.dtype.kind == "V"` to see if a structured `dtype` is present.

**Comment by sjperkins** [comment]
I suspect the memory issues are similar to those described in https://github.com/dask/dask/issues/874, as unique creates a single output node for the values (similar to the mean computation in https://github.com/dask/dask/issues/874)
